### DISCUSSION TITLE
Any plans for astexplorer or more examples for ast javascript/typescript api?

### DISCUSSION THREAD

**Discussion by aelbore** [swc-project/swc#1677] [Ideas]
most of my typescript transformer are using typescript, it would be good to have more example of ast explorer or more example of ast visitor.

**Comment by kdy1** [comment]
I expect an AST explorer to be super useful. It's simply me not having time to do it.

**Reply by mischnic** [reply]
It's very easy to add a new parser to astexplorer.net (see below), but a problem that I ran into was that the spans returned by `swc.parseSync` (via `@swc/wasm-web`) only contains these somewhat arbitrary start + end numbers. What astexplorer needs to make hover-highlighting work is a way to convert that `span` object into a range that is a start/end string index into the source string. Is there a way to achieve that? I couldn't find anything exposed in https://github.com/swc-project/swc/blob/master/node-swc/src/index.ts


<img width="1228" alt="Bildschirmfoto 2021-05-12 um 01 49 01" src="https://user-images.githubusercontent.com/4586894/117898001-3ce66b80-b2c4-11eb-8008-dee5db2488db.png">

**Reply by aelbore** [reply]
@kdy1 thanks for your hardwork in swc, @mischnic thanks this super nice so `swc.parse` or `swc.parseSync` return `ast`?

**Reply by mischnic** [reply]
Yes. I've pushed my changes here: https://github.com/mischnic/astexplorer/commit/bb113f9

It works fine apart from the hover highlighting. You can run it locally by following the instructions in the readme.

**Reply by aelbore** [reply]
@mischnic I clone your repository branch swc i got some error , im using ubuntu subsystem in windows 
any thanks ^^
![image](https://user-images.githubusercontent.com/18069807/118140878-5f4ec500-b43b-11eb-8239-5ff228431cd5.png)

**Reply by jondlm** [reply]
[→ Minimal repro of the span numbering issue](https://jondlm.github.io/swc-wasm-web-span-issue/) ([code](https://github.com/jondlm/swc-wasm-web-span-issue))

I spent a bunch of time today digging into why the `span`s are funky in ast explorer but not the swc playground. For the life of me I could not figure out why it's busted in ast explorer. For some reason running `parseSync` multiple times causes the span numbers to compound upon each other making them useless after the first render. It seems like something to do with how the Rust code is managing memory. Even the first render's span numbers are off by 1 for some reason only outside of the swc playground.

My current hunch is that somehow the way swc playground uses [swr](https://swr.vercel.app/) is changing the way the wasm memory is handled. It's a far fetched theory but it's the best I've got at the moment.

**Reply by morganney** [reply]
@jondlm https://github.com/knightedcodemonkey/reparse

Workaround for https://github.com/swc-project/swc/issues/1366. Try running in a web worker.
